util/unifyfs-stage/src/unifyfs-stage-transfer.c

```diff
@@ -0,0 +1,376 @@
+    }
+
+out:
+    /* MD5_xx returns 1 for success */
+    ret = ret == 1 ? 0 : EIO;
+    close(fd);
+
+    return ret;
+}
+
+static inline char* checksum_str(char* buf, unsigned char* digest)
+{
+    int i = 0;
+    char* pos = buf;
+
+    for (i = 0; i < MD5_DIGEST_LENGTH; i++) {
+        pos += sprintf(pos, "%02x", digest[i]);
+    }
+
+    pos[0] = '\0';
+
+    return buf;
+}
+
+static int verify_checksum(const char* src, const char* dst)
+{
+    int ret = 0;
+    int i = 0;
+    char md5src[2 * MD5_DIGEST_LENGTH + 1] = { 0, };
+    char md5dst[2 * MD5_DIGEST_LENGTH + 1] = { 0, };
+    unsigned char src_digest[MD5_DIGEST_LENGTH + 1] = { 0, };
+    unsigned char dst_digest[MD5_DIGEST_LENGTH + 1] = { 0, };
+
+    src_digest[MD5_DIGEST_LENGTH] = '\0';
+    dst_digest[MD5_DIGEST_LENGTH] = '\0';
+
+    ret = md5_checksum(src, src_digest);
+    if (ret) {
+        fprintf(stderr, "failed to calculate checksum for %s (%s)\n",
+                        src, strerror(ret));
+        return ret;
+    }
+
+    ret = md5_checksum(dst, dst_digest);
+    if (ret) {
+        fprintf(stderr, "failed to calculate checksum for %s (%s)\n",
+                        dst, strerror(ret));
+        return ret;
+    }
+
+    if (verbose) {
+        printf("[%d] src: %s, dst: %s\n", rank,
+               checksum_str(md5src, src_digest),
+               checksum_str(md5dst, dst_digest));
+    }
+
+    for (i = 0; i < MD5_DIGEST_LENGTH; i++) {
+        if (src_digest[i] != dst_digest[i]) {
+            fprintf(stderr, "[%d] checksum verification failed: "
+                            "(src=%s, dst=%s)\n", rank,
+                            checksum_str(md5src, src_digest),
+                            checksum_str(md5dst, dst_digest));
+            ret = EIO;
+        }
+    }
+
+    return ret;
+}
+
+static inline char* str_rtrim(char* str)
+{
+    if (str) {
+        char* pos = &str[strlen(str) - 1];
+
+        while (pos >= str && isspace(*pos)) {
+            *pos = '\0';
+            pos--;
+        }
+    }
+
+    return str;
+}
+
+// format of the manifest file lines (for the moment) is as
+// follows: each line consists of the source filename including
+// full path, then one or more spaces, and then the destination
+// filename including full path.  If neither of the filenames
+// contain spaces, the filenames can lack quotes around them.
+// If the filenames contain spaces, then they may be surrounded
+// with double quote symbols (").  The only characters that are
+// disallowed with this format are the double-quote symbol
+// and the newline character.
+static inline int unifyfs_parse_manifest_line
+                  (char* line, char** src, char** dst)
+{
+    char* ptr_doublequotes[4];
+    char* CSP = NULL; // "current search pointer"
+    char* loc_of_space = NULL;
+    char* loc_of_newline = NULL;
+    char test_char;
+    size_t offset;
+
+    // does the first filename start with a doublequote character?
+    if ((*line) == '"') {
```

Comment:
Did you consider using `strtok()` to get the quoted filenames (if any)?

Reply:
I hadn't.  It might be suitable, but I find the man page a bit murky to understand completely.  

I will need to do a bit of experimentation before I'm confident I can invoke strtok() correctly and consistently.  I'm sure we'll want to do a cleanup of the manifest line parsing in the near future.  We can add strtok() then.

Reply:
`strtok()` is the way to go. Learn it, love it, and stop reinventing libc functions.

Reply:
I hadn't used strtok() before.  Whenever possible I parse lines using regex in perl, and I was in a hurry.  

If you still want this change before approving I'll make this change last.

Reply:
Please use this simplified manifest parser:
https://gist.github.com/tonyhutter/135155fc185f568d61c046875d39699c
It allows quoted filenames, a whitespace separator (not just space), and ignores trailing/leading whitespace.  This allows the user wants to indent certain lines, or tab-align source/destination filenames if they want.